db/virtual_table.cc

```diff
@@ -106,4 +106,91 @@ mutation_source memtable_filling_virtual_table::as_mutation_source() {
     });
 }
 
+mutation_source streaming_virtual_table::as_mutation_source() {
+    return mutation_source([this] (schema_ptr s,
+        reader_permit permit,
+        const dht::partition_range& pr,
+        const query::partition_slice& slice,
+        const io_priority_class& pc,
+        tracing::trace_state_ptr trace_state,
+        streamed_mutation::forwarding fwd,
+        mutation_reader::forwarding fwd_mr) {
+
+        if (fwd == streamed_mutation::forwarding::yes) {
+            throw std::runtime_error("streamed_mutation::forwarding::yes unsupported in virtual tables");
+        }
+
+        if (fwd_mr == mutation_reader::forwarding::yes) {
+            throw std::runtime_error("mutation_reader::forwarding::yes unsupported in virtual tables");
+        }
+
+        // We cannot pass the partition_range directly to execute()
+        // because it is not guaranteed to be alive until execute() resolves.
+        // It is only guaranteed to be alive as long as the returned reader is alive.
+        // We achieve safety by mediating access through query_restrictions. When the reader
+        // dies, pr is cleared and execute() will get an exception.
+        struct my_result_collector : public result_collector, public query_restrictions {
+            queue_reader_handle handle;
+
+            // Valid until handle.is_terminated(), which is set to true when the
+            // queue_reader dies.
+            const dht::partition_range* pr;
+            mutation_reader::forwarding fwd_mr;
+
+            my_result_collector(schema_ptr s, reader_permit p, const dht::partition_range* pr, queue_reader_handle&& handle)
+                : result_collector(s, p)
+                , handle(std::move(handle))
+                , pr(pr)
+            { }
+
+            // result_collector
+            future<> take(mutation_fragment fragment) override {
+                return handle.push(std::move(fragment));
+            }
+
+            // query_restrictions
+            const dht::partition_range& partition_range() const override {
+                if (handle.is_terminated()) {
+                    throw std::runtime_error("read abandoned");
+                }
+                return *pr;
+            }
+        };
+
+        // FIXME: queue_reader doesn't support next_partition() needed for "distinct".
```

Comment:
As `next_partition()` now returns a `future<>` an implementation can easily be added.

Reply:
Done, but I'm not sure it's the proper way.